daemon/api_themes.go

```diff
@@ -0,0 +1,251 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/overlord"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/store"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	themesCmd = &Command{
+		Path:   "/v2/accessories/themes",
+		UserOK: true,
+		GET:    checkThemes,
+		POST:   installThemes,
+	}
+)
+
+type themeStatus string
+
+const (
+	themeInstalled   themeStatus = "installed"
+	themeAvailable   themeStatus = "available"
+	themeUnavailable themeStatus = "unavailable"
+)
+
+type themeStatusResponse struct {
+	GtkThemes   map[string]themeStatus `json:"gtk-themes"`
+	IconThemes  map[string]themeStatus `json:"icon-themes"`
+	SoundThemes map[string]themeStatus `json:"sound-themes"`
+}
+
+func installedThemes(overlord *overlord.Overlord) (gtkThemes, iconThemes, soundThemes []string, err error) {
+	infos := overlord.InterfaceManager().Repository().Info(&interfaces.InfoOptions{
+		Names: []string{"content"},
+		Slots: true,
+	})
+	for _, info := range infos {
+		for _, slot := range info.Slots {
+			var content string
+			// The content interface ensures this attribute exists
+			if err := slot.Attr("content", &content); err != nil {
+				return nil, nil, nil, err
+			}
+			var themes *[]string
+			switch content {
+			case "gtk-3-themes":
+				themes = &gtkThemes
+			case "icon-themes":
+				themes = &iconThemes
+			case "sound-themes":
+				themes = &soundThemes
+			default:
+				continue
+			}
+			var sources []interface{}
+			if err := slot.Attr("source.read", &sources); err != nil {
+				continue
+			}
+			for _, s := range sources {
+				if path, ok := s.(string); ok {
+					*themes = append(*themes, filepath.Base(path))
+				}
+			}
+		}
+	}
+	sort.Strings(gtkThemes)
+	sort.Strings(iconThemes)
+	sort.Strings(soundThemes)
+	return gtkThemes, iconThemes, soundThemes, nil
+}
+
+var badPkgCharRegexp = regexp.MustCompile(`[^a-z0-9]+`)
+
+func themePackageCandidates(prefix, themeName string) []string {
+	themeName = strings.ToLower(themeName)
+	themeName = badPkgCharRegexp.ReplaceAllString(themeName, "-")
+	themeName = strings.Trim(themeName, "-")
+
+	var packages []string
+	for themeName != "" {
+		packages = append(packages, prefix+themeName)
+		pos := strings.LastIndexByte(themeName, '-')
+		if pos < 0 {
+			break
+		}
+		themeName = themeName[:pos]
+	}
+	return packages
+}
+
+func collectThemeStatusForPrefix(ctx context.Context, theStore snapstate.StoreService, user *auth.UserState, prefix string, themes, installed []string, status map[string]themeStatus, candidateSnaps map[string]bool) error {
```

Comment:
(nitpick^2) maybe a personal preference but I would prefer if we could make it clearer here what arguments will be mutated and what are pure inputs.Ideally only have inputs and return data. Not sure how well this fits though. Anyway, not a blocker but reading this it's a bit unclear to me without reading this very carefully what the data flow is (but maybe just me).

(edit) I played around with this and I'm not sure the extra complexity is worth it. Anyway, not a blocker.
```
diff --git a/daemon/api_themes.go b/daemon/api_themes.go
index e86cf72e7d..3ec69d72ed 100644
--- a/daemon/api_themes.go
+++ b/daemon/api_themes.go
@@ -122,7 +122,9 @@ func themePackageCandidates(prefix, themeName string) []string {
        return packages
 }
 
-func collectThemeStatusForPrefix(ctx context.Context, theStore snapstate.StoreService, user *auth.UserState, prefix string, themes, installed []string, status map[string]themeStatus, candidateSnaps map[string]bool) error {
+func collectThemeStatusForPrefix(ctx context.Context, theStore snapstate.StoreService, user *auth.UserState, prefix string, themes, installed []string, status map[string]themeStatus) (candidateSnaps map[string]bool, err error) {
+       candidateSnaps = make(map[string]bool)
+
        for _, theme := range themes {
                // Skip duplicates
                if _, ok := status[theme]; ok {
@@ -139,7 +141,7 @@ func collectThemeStatusForPrefix(ctx context.Context, theStore snapstate.StoreSe
                        if info, err = theStore.SnapInfo(ctx, store.SnapSpec{Name: name}, user); err == store.ErrSnapNotFound {
                                continue
                        } else if err != nil {
-                               return err
+                               return nil, err
                        }
                        // Only mark the theme as available if it has
                        // been published to the stable channel.
@@ -150,7 +152,7 @@ func collectThemeStatusForPrefix(ctx context.Context, theStore snapstate.StoreSe
                        }
                }
        }
-       return nil
+       return candidateSnaps, nil
 }
 
 func themeStatusAndCandidateSnaps(ctx context.Context, c *Command, user *auth.UserState, gtkThemes, iconThemes, soundThemes []string) (status themeStatusResponse, candidateSnaps map[string]bool, err error) {
@@ -163,18 +165,29 @@ func themeStatusAndCandidateSnaps(ctx context.Context, c *Command, user *auth.Us
        status.GtkThemes = make(map[string]themeStatus, len(gtkThemes))
        status.IconThemes = make(map[string]themeStatus, len(iconThemes))
        status.SoundThemes = make(map[string]themeStatus, len(soundThemes))
-       candidateSnaps = make(map[string]bool)
-       if err = collectThemeStatusForPrefix(ctx, theStore, user, "gtk-theme-", gtkThemes, installedGtk, status.GtkThemes, candidateSnaps); err != nil {
+       candidatesGtk, err := collectThemeStatusForPrefix(ctx, theStore, user, "gtk-theme-", gtkThemes, installedGtk, status.GtkThemes)
+       if err != nil {
                return themeStatusResponse{}, nil, err
        }
-       if err = collectThemeStatusForPrefix(ctx, theStore, user, "icon-theme-", iconThemes, installedIcon, status.IconThemes, candidateSnaps); err != nil {
+       candidatesIcon, err := collectThemeStatusForPrefix(ctx, theStore, user, "icon-theme-", iconThemes, installedIcon, status.IconThemes)
+       if err != nil {
                return themeStatusResponse{}, nil, err
        }
-       if err = collectThemeStatusForPrefix(ctx, theStore, user, "sound-theme-", soundThemes, installedSound, status.SoundThemes, candidateSnaps); err != nil {
+       candidatesSound, err := collectThemeStatusForPrefix(ctx, theStore, user, "sound-theme-", soundThemes, installedSound, status.SoundThemes)
+       if err != nil {
                return themeStatusResponse{}, nil, err
        }
+       merged := func(ml ...map[string]bool) map[string]bool {
+               merged := make(map[string]bool)
+               for _, m := range ml {
+                       for k, v := range m {
+                               merged[k] = v
+                       }
+               }
+               return merged
+       }
 
-       return status, candidateSnaps, nil
+       return status, merged(candidatesGtk, candidatesIcon, candidatesSound), nil
 }
 
 func checkThemes(c *Command, r *http.Request, user *auth.UserState) Response {
diff --git a/daemon/api_themes_test.go b/daemon/api_themes_test.go
index e8f39fa37a..cfcf4759d2 100644
--- a/daemon/api_themes_test.go
+++ b/daemon/api_themes_test.go
@@ -169,9 +169,8 @@ func (s *themesSuite) TestThemeStatusForPrefix(c *C) {
 
        ctx := context.Background()
        status := make(map[string]daemon.ThemeStatus)
-       toInstall := make(map[string]bool)
 
-       err := daemon.CollectThemeStatusForPrefix(ctx, s, nil, "gtk-theme-", []string{"Installed", "Installed", "Available", "Unavailable"}, []string{"Installed"}, status, toInstall)
+       toInstall, err := daemon.CollectThemeStatusForPrefix(ctx, s, nil, "gtk-theme-", []string{"Installed", "Installed", "Available", "Unavailable"}, []string{"Installed"}, status)
        c.Check(err, IsNil)
        c.Check(status, DeepEquals, map[string]daemon.ThemeStatus{
                "Installed":   daemon.ThemeInstalled,
@@ -196,9 +195,8 @@ func (s *themesSuite) TestThemeStatusForPrefixStripsSuffixes(c *C) {
 
        ctx := context.Background()
        status := make(map[string]daemon.ThemeStatus)
-       toInstall := make(map[string]bool)
 
-       err := daemon.CollectThemeStatusForPrefix(ctx, s, nil, "gtk-theme-", []string{"Yaru-dark"}, nil, status, toInstall)
+       toInstall, err := daemon.CollectThemeStatusForPrefix(ctx, s, nil, "gtk-theme-", []string{"Yaru-dark"}, nil, status)
        c.Check(err, IsNil)
        c.Check(status, DeepEquals, map[string]daemon.ThemeStatus{
                "Yaru-dark": daemon.ThemeAvailable,
@@ -221,9 +219,8 @@ func (s *themesSuite) TestThemeStatusForPrefixIgnoresUnstable(c *C) {
 
        ctx := context.Background()
        status := make(map[string]daemon.ThemeStatus)
-       toInstall := make(map[string]bool)
 
-       err := daemon.CollectThemeStatusForPrefix(ctx, s, nil, "gtk-theme-", []string{"Yaru"}, nil, status, toInstall)
+       toInstall, err := daemon.CollectThemeStatusForPrefix(ctx, s, nil, "gtk-theme-", []string{"Yaru"}, nil, status)
        c.Check(err, IsNil)
        c.Check(status, DeepEquals, map[string]daemon.ThemeStatus{
                "Yaru": daemon.ThemeUnavailable,
@@ -238,9 +235,8 @@ func (s *themesSuite) TestThemeStatusForPrefixReturnsErrors(c *C) {
 
        ctx := context.Background()
        status := make(map[string]daemon.ThemeStatus)
-       toInstall := make(map[string]bool)
 
-       err := daemon.CollectThemeStatusForPrefix(ctx, s, nil, "gtk-theme-", []string{"Theme"}, nil, status, toInstall)
+       toInstall, err := daemon.CollectThemeStatusForPrefix(ctx, s, nil, "gtk-theme-", []string{"Theme"}, nil, status)
        c.Check(err, Equals, s.err)
        c.Check(status, DeepEquals, map[string]daemon.ThemeStatus{
                "Theme": daemon.ThemeUnavailable,
```

Reply:
Oh, I see in the history that this was discussed so probably redundant.